apps/client/src/features/app-settings/panel/project-panel/ProjectList.tsx

```diff
@@ -1,19 +1,42 @@
-import { IconButton, Menu, MenuButton, MenuItem, MenuList } from '@chakra-ui/react';
-import { IoEllipsisHorizontal } from '@react-icons/all-files/io5/IoEllipsisHorizontal';
+import { useMemo, useState } from 'react';
 
 import { useProjectList } from '../../../../common/hooks-query/useProjectList';
 import * as Panel from '../PanelUtils';
 
-import style from './ProjectPanel.module.scss';
+import ProjectListItem from './ProjectListItem';
+
+export type EditMode = 'rename' | 'duplicate' | null;
 
 export default function ProjectList() {
-  const { data } = useProjectList();
+  const { data, refetch } = useProjectList();
   const { files, lastLoadedProject } = data;
 
-  // extract currently loaded from file list
-  const currentlyLoadedIndex = files.findIndex((project) => project.filename === lastLoadedProject);
-  const projectFiles = [...files];
-  const current = projectFiles.splice(currentlyLoadedIndex, 1)[0];
+  const [editingMode, setEditingMode] = useState<EditMode | null>(null);
+  const [editingFilename, setEditingFilename] = useState<string | null>(null);
+
+  const handleToggleEditMode = (editMode: EditMode, filename: string | null) => {
+    setEditingMode((prev) => (prev === editMode && filename === editingFilename ? null : editMode));
+    setEditingFilename(filename);
+  };
+
+  const handleClear = () => {
+    setEditingMode(null);
+    setEditingFilename(null);
+  };
+
+  const handleRefetch = async () => {
+    await refetch();
+  };
+
+  const reorderedProjectFiles = useMemo(() => {
+    if (!data?.files?.length) return [];
+
+    const currentlyLoadedIndex = files.findIndex((project) => project.filename === lastLoadedProject);
+    const projectFiles = [...files];
+    const current = projectFiles.splice(currentlyLoadedIndex, 1)?.[0];
+
+    return [current, ...projectFiles];
+  }, [data?.files?.length, files, lastLoadedProject]);
```

Comment:
Question: since we pass the placeholder in the `useProjectList` hook, data will never be nullish, so I dont think we need the optional chaining?